docs/src/tutorial.md

```diff
@@ -0,0 +1,248 @@
+  giving
+
+## ADNLPModel Tutorial
+
+ADNLPModel is very simple to use and is very useful for classrooms, for
+instance.
+It only needs the objective function $f$ and a starting point $x^0$ to be
+well-defined.
+For constrained problems, you'll also need the constraints function $c$, and
+the constraints vectors $c_L$ and $c_U$, such that $c_L \leq c(x) \leq c_U$.
+Equality constraints are identified by $c_{L_i} = c_{U_i}$.
+
+Let's define the famous Rosenbrock function
+\begin{align*}
+f(x) = (x_1 - 1)^2 + 100(x_2 - x_1^2)^2,
+\end{align*}
+with starting point $x^0 = (-1.2,1.0)$.
+
+```@example adnlp
+using NLPModels
+
+f(x) = (x[1] - 1.0)^2 + 100*(x[2] - x[1]^2)^2
+x0 = [-1.2; 1.0]
+nlp = ADNLPModel(f, x0)
+```
+
+This is enough to define the model.
+Let's get the objective function value at $x^0$, using only `nlp`.
+
+```@example adnlp
+fx = obj(nlp, nlp.meta.x0)
+println("fx = $fx")
+```
+
+Done.
+Let's try the gradient and Hessian.
+
+```@example adnlp
+gx = grad(nlp, nlp.meta.x0)
+Hx = hess(nlp, nlp.meta.x0)
+println("gx = $gx")
+println("Hx = $Hx")
+```
+
+Notice how only the lower triangle of the Hessian is stored.
+Also notice that it is *dense*. This is a current limitation of this model. It
+doesn't return sparse matrices, so use it with care.
+
+Let's do something a little more complex here, defining a function to try to
+solve this problem through gradient method with Armijo search.
+Namely, the method
+
+1. Given $x^0$, $\varepsilon > 0$, and $\eta \in (0,1)$. Set $k = 0$;
+2. Compute $d^k = -\nabla f(x^k)$;
+3. Compute $\alpha_k$ such that
+$ f(x^k + \alpha_kd^k) < f(x^k) + \alpha_k\eta \nabla f(x^k)^Td^k $
+4. Define $x^{k+1} = x^k + \alpha_kx^k$
+5. Update $k = k + 1$
+6. If $\Vert \nabla f(x^k) \Vert < \varepsilon$ STOP with $x^* = x^k$,
+otherwise go to step 2.
+
+```@example adnlp
+function gradient(nlp; itmax=100000, eta=1e-2, eps=1e-6, sigma=0.9)
+  x = nlp.meta.x0
+  fx = obj(nlp, x)
+  gx = grad(nlp, x)
+  gtg = dot(gx, gx)
+  ef = 0
+  iter = 0
+  while gtg > eps^2
+    t = 1.0
+    while obj(nlp, x - t*gx) > fx - eta*t*gtg
+      t *= sigma
+    end
+    x = x - t*gx
+    fx = obj(nlp, x)
+    gx = grad(nlp, x)
+    gtg = dot(gx, gx)
+    iter += 1
+    if iter >= itmax
+      ef = 1
+      break
+    end
+  end
+  return x, fx, sqrt(gtg), ef, iter
+end
+
+x, fx, ngx, ef, iter = gradient(nlp)
+println("x = $x")
+println("fx = $fx")
+println("ngx = $ngx")
+println("ef = $ef")
+println("iter = $iter")
+```
+
+Maybe this code is too complicated? If you're in a class you just want to show a
+Newton step.
+
+```@example adnlp
+g(x) = grad(nlp, x)
+H(x) = hess(nlp, x) + triu(hess(nlp, x)', 1)
+x = nlp.meta.x0
+d = -H(x)\g(x)
+```
+
+or a few
+
+```@example adnlp
+for i = 1:5
+  x = x - H(x)\g(x)
+  println("x = $x")
+end
+```
+
+Also, notice how we can reuse the method.
+
+```@example adnlp
+f(x) = (x[1] + x[2] - 4)^2 + (x[1]*x[2] - 1)^2
+x0 = [2.0; 1.0]
+nlp = ADNLPModel(f, x0)
+
+x, fx, ngx, ef, iter = gradient(nlp)
```

Comment:
Let's call the method `steepest` or something like that to avoid the possible confusion that `gradient` computes a gradient.